feat(hardware-testing, shared-data): DVT pipettes in shared-data; HW assembly qc test script

andySigler opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

This PR adds `pipette_assembly_qc_ot3` script to hardware-testing, and makes some other changes to facilitate that script.

NOTE: have not yet tested on an unmodified DVT pipette. But this script works and was tested on an EVT ot3 with:
- EVT single-channel pipettes
- EVT pipettes with DVT serial numbers
- electrically modified DVT pipettes

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

- adds `v3.4` model pipette to shared data
- fixes bug in shared-data where pick-up-tip-speed was too high (`10` instead of `5`)
- adds script `hardware_testing.production_qc.pipette_assembly_qc_ot3` to test fully assembled DVT pipettes
- adds script `hardware_testing.production_qc.pipette_current_speed_qc_ot3` to test partially assembled DVT pipettes
  - this script is ran on a fixture, not a fully assembled OT3, that is why it does not use `OT3API` at the moment
- adds pressure-fixture Arduino firmware source files to `./opentrons/hardware-testing/firmware` folder
- adds pressure-fixture driver

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

Carlos-fernandez committed:
production_multis_v3.3

Carlos-fernandez committed:
few fixes to the test script

Carlos-fernandez committed:
working assembly qc test script

Carlos-fernandez committed:
ot3 pressure fixture driver

Carlos-fernandez committed:
change comment

Carlos-fernandez committed:
init file for pressure fixture

Carlos-fernandez committed:
multi_channel_qc_test updates

Carlos-fernandez committed:
some conflict fixes

Carlos-fernandez committed:
format fixes

Carlos-fernandez committed:
Merge branch 'edge' into production_multis_v3.3

Carlos-fernandez committed:
changed name to pipette_qc_test

Carlos-fernandez committed:
Changed description

Carlos-fernandez committed:
fixed lint errors on ot3 driver

Carlos-fernandez committed:
a few more fixes

andySigler committed:
Merge branch 'edge' into production_multis_v3.3

andySigler committed:
WIP: adds firmware files for 8ch pressure sensor fixture

andySigler committed:
revert back to original from Peter Li

andySigler committed:
adds version to firmware

andySigler committed:
adds small test to bottom of pressure-fixture driver

andySigler committed:
adds left/right configuration for channels

andySigler committed:
adds method to read pipette serial number

andySigler committed:
wip: cleaning things up

andySigler committed:
assembly qc script saves data

andySigler committed:
linting!

andySigler committed:
record 10 samples of pressure data to csv

andySigler committed:
wip: trying to get fw-update working again

andySigler committed:
Merge branch 'edge' into production_multis_v3.3

andySigler committed:
reduce pick-up-current for single-channels to 5 mm/sec

andySigler committed:
test runs and saves data

andySigler committed:
display final test results at end of script

andySigler committed:
moves pressure-fixture-event things to separate submodule

andySigler committed:
pull from new testing-data directory

andySigler committed:
adds plunger test

andySigler committed:
diagnostics tests working

andySigler committed:
Merge branch 'edge' into production_multis_v3.3

andySigler committed:
log all data and put results at top of file

andySigler committed:
remove some jogging messages

andySigler committed:
allows for different aspirate volumes and pressure configs for p50 and p1000

andySigler committed:
average cap readings; use tips down column

andySigler committed:
stall plunger using speed instead of collision

andySigler committed:
avoid over-pressurizing sensors

andySigler committed:
more setting tunings; delete-data command works

andySigler committed:
firmware prints values to 5 decimal places

andySigler committed:
adds header to pressure-data

andySigler committed:
firmware uses Pascals not kPa

andySigler committed:
adds current/speed test using opentrons_hardware

andySigler committed:
adds DVT version of pipettes

andySigler committed:
Merge branch 'edge' into feat-hardware-testing-assembly_qc_single_channel_dvt

andySigler committed:
wip: linting

andySigler committed:
Merge branch 'feat-hardware-testing-assembly_qc_single_channel_dvt' of https://github.com/Opentrons/opentrons into feat-hardware-testing-assembly_qc_single_channel_dvt

andySigler committed:
wip: slowing down gantry to avoid stall detection bug

andySigler committed:
wip

andySigler committed:
wip: linting

andySigler committed:
Merge branch 'edge' into feat-hardware-testing-assembly_qc_single_channel_dvt

andySigler committed:
reset data arrays before each test

andySigler committed:
add DVT minor version to shared-data python package

andySigler committed:
change simulated pipette models to DVT version

andySigler committed:
pressure-data header with nothing in first row

andySigler committed:
adds some dashed lines at top, to make copy/paste easier

andySigler committed:
format top line like how Starno wants

andySigler committed:
increase flake8 max-complexity from 9 to 20 b/c test scripts handle all logic in main()

andySigler committed:
linting

Laura-Danielle commented on hardware-testing/hardware_testing/production_qc/pipette_current_speed_qc_ot3.py:
```diff
@@ -0,0 +1,225 @@
+
+DEFAULT_CURRENT_SEQUENCE = [0.6, 0.5, 0.4, 0.3, 0.2, 0.15, 0.1, 0.05]
+CUSTOM_CURRENT_SEQUENCE = {
+    "p1000_single_v43": None,
```
Is it possible to stop using v3 versus v4 to denote which tip size we're using? Otherwise it's gonna get real confusing. We already support different tip functions for the same pipette model.

andySigler replied:
I removed the line from the `pipette_current_speed_qc_ot3.py` script, but I'm not quite sure what your requesting re: shared-data. I don't see pipette model versions specified in the `pipette/definitions/2/**`, so I don't see what I should be mimicking.

Laura-Danielle reviewed (changes requested):
If we want to differentiate between `v3_3` and `v3_4`, we should add an additional version called `3_4.json` (in `pipette/definitions/2/**`). Once I add GEN1/GEN2 pipettes to the new configuration then we will fully support versioning in the v2 schema version of pipettes.

Would be _great_ to not have to keep adding to v1 of the schema version for pipette QC. Let us know what we can do to support that transition.

sfoster1 committed:
hack(hardware)!: hardcode left pipette presence (#12002)

After https://github.com/Opentrons/ot3-firmware/pull/544, which is
required because of hardware changes and short time to adapt, the head
is (temporarily) no longer capable of detecting the presence of
pipettes. This is a problem for the pipette QC script, which uses the
mainline pipette detection which relies on the head. Temporarily work
around this by hardcoding the python side to always return that a
pipette is present on the left mount.

This will make the script only work for a pipette on the left mount.

This may be merged into the working branch for DVT, but it MUST NOT be
merged in to edge, ever.

andySigler committed:
adds a few comment to qc script for future improvements

andySigler committed:
remove v4.3 from current/speed test

sfoster1 commented on hardware/opentrons_hardware/hardware_control/tools/detector.py:
```diff
@@ -163,11 +163,7 @@ async def _handle_pipette_info(
     async def detect(self, timeout_sec: float = 1.0) -> ToolSummary:
         """Run once and detect tools."""
         with WaitableCallback(self._messenger) as wc:
-            attached_status_request = message_definitions.AttachedToolsRequest()
```
Note for future: Remove this a) before merging b) once https://github.com/Opentrons/ot3-firmware/pull/551 is merged c) once all hardware is DVT or has its mount id stuff reworked as mentioned in that pr

sfoster1 reviewed (changes requested):
extra request-changes to prevent merging and not what has to be changed before merge (branch currently has a hack to work around DVT pipette mount detection not working)

andySigler committed:
adds user input to enter room's celsius/humidity

andySigler committed:
merge in latest changes

andySigler committed:
merge in edge

andySigler committed:
undo hack to only use left mount

andySigler committed:
merge in edge

andySigler committed:
remove applying addition offset to already instantiated labware object

andySigler committed:
Merge branch 'edge' into feat-hardware-testing-assembly_qc_single_channel_dvt

andySigler committed:
adds 3_4.json to pipette definitions 2

andySigler committed:
adds DVT multis to shared data

andySigler committed:
linting

Laura-Danielle reviewed (approved):
This lgtm now.

andySigler committed:
fix merge conflicts

sfoster1 commented on shared-data/pipette/definitions/1/pipetteModelSpecs.json:
```diff
@@ -6128,6 +6128,201 @@
       "returnTipHeight": 0.83,
       "idleCurrent": 0.3
     },
+    "p1000_single_v3.4": {
```
@Laura-Danielle do we need this entry?

Laura-Danielle replied:
We do, for the QC fixture test on the OT2. Right now, the OT2 has not been switched over to look up from V2 version of the pip configs. We need to first support mutable configs in the new format to not break that feature in the app 😢

sfoster1 reviewed (approved):
- that `hardware-testing/build_fw.sh` is empty
- not sure if we need the new entry in the schema v1 pipette defs

Other than that looks good and won't block on those since they're not _wrong_, per se
